optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/util/ConcurrentCache.java

```diff
@@ -0,0 +1,31 @@
+package org.optaplanner.core.impl.domain.solution.util;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+
+/**
+ * Thread-safe basic cache implementation.
+ *
+ * @param <K> the type of keys for the cache
+ * @param <V> the type of values stored in the cache
+ * */
+public final class ConcurrentCache<K, V> extends ConcurrentHashMap<K, V> {
+
+    /**
+     * {@link ConcurrentHashMap#computeIfAbsent(Object, Function)} should not be used to directly retrieve cached values
+     * as it incorporates locking even when the value is present and doesn't have to be computed. Thus, the value is
+     * retrieved by {@link ConcurrentHashMap#get(Object)} and if and only if the value is null,
+     * {@link ConcurrentHashMap#computeIfAbsent(Object, Function)} is used to ensure thread-safety.
+     *
+     * {@inheritDoc}
+     * */
+    @Override
+    public V computeIfAbsent(K key, Function<? super K, ? extends V> mappingFunction) {
+        V value = get(key);
+        if (value == null) {
+            value = super.computeIfAbsent(key, mappingFunction);
+        }
+        return value;
```

Comment:
This looks a lot like the (broken) double locked idiom:
  https://www.javaworld.com/article/2074979/java-concurrency/double-checked-locking--clever--but-broken.html

What happens to performance if we make this method synchronized?

Reply:
In this case, I think it doesn't matter. The cache update is properly synchronized - the worst thing we're risking is that the computation will happen twice. Once one thread stores the result, every other thread will see the cached result.

Reply:
We are not testing and updating a shared field - `value` is a local variable. If it's null and two threads will end up calling `super.computeIfAbsent()`, they will get synchronized inside the `computeIfAbsent` method, so as Lukas pointed out, both thread will compute the value in this situation. If the computed value depends on what thread executes it (which it shouldn't), the last thread wins.

Reply:
Agreed, good point. Because value is a local variable it is not a double checked locking.

Reply:
Let me just correct myself, after looking with @yurloc into the ConcurrentHashMap implementation more carefully, even the computation cannot happen twice - the collection guarantees that.

Reply:
DO NOT MERGE yet - I would prefer running complete benchmarks first to make sure the fix addresses all the performance issues found.

Reply:
I was just about to hit the merge button :)
I tried your approach locally and I can confirm that the performance notecably improves on the micro benchmark.
I think it's ready to merge, especially on master. I might nitpick on a few lines. I 'll then backport it to 7.5.x once it's merged.